registry/auth/token/token.go

```diff
@@ -172,26 +172,27 @@ func (t *Token) VerifySigningKey(verifyOpts VerifyOptions) (signingKey crypto.Pu
 	// verifying the first one in the list only at the moment.
 	header := t.JWT.Headers[0]
 
-	signingKey, err = verifyCertChain(header, verifyOpts.Roots)
-	// NOTE(milosgajdos): if the x5c header is missing
-	// the token may have been signed by a JWKS.
-	if err != nil && err != jose.ErrMissingX5cHeader {
-		return
-	}
-
-	switch {
-	case header.JSONWebKey != nil:
-		signingKey, err = verifyJWK(header, verifyOpts)
-	case len(header.KeyID) > 0:
-		signingKey = verifyOpts.TrustedKeys[header.KeyID]
-		if signingKey == nil {
-			err = fmt.Errorf("token signed by untrusted key with ID: %q", header.KeyID)
+	signingKey, err := verifyCertChain(header, verifyOpts.Roots)
+	if err != nil {
+		// NOTE(milosgajdos): if the x5c header is missing
+		// the token may have been signed by a JWKS.
+		if err == jose.ErrMissingX5cHeader {
+			switch {
+			case header.JSONWebKey != nil:
+				return verifyJWK(header, verifyOpts)
+			case len(header.KeyID) > 0:
+				if signingKey, ok := verifyOpts.TrustedKeys[header.KeyID]; ok {
+					return signingKey, nil
+				}
+				return nil, fmt.Errorf("token signed by untrusted key with ID: %q", header.KeyID)
```

Comment:
Not new, but I notice we return a "ErrInvalidToken" in other branches; is that error special / used as a sentinel error? (and in that case, must all invalid errors wrap that one somewhere?)

Reply:
`ErrinvalidToken` is a sentinel error that was in the original code (see v2.8 branch: https://github.com/distribution/distribution/blob/5d5c60f77c01b143b22be9d62af3b32843fbcb0c/registry/auth/token/token.go#L31)

I think in this specific code we _could_ maybe wrap it, but I'm not sure it's that important, because the other `switch` branches return different errors, too 🤔

Reply:
I am also wondering if we should ditch the `switch` clause here and replace it with `if` statements 🤔 might make the code a bit more readble

Reply:
Nah, I think a switch is nice; with a switch, all conditions are upfront and visible, with `if/else if/else`, they get buried deeper down the line, which (IMO) usually makes it _less_ readable.

Reply:
Anything else that needs changing?